lib/recursive-methods.rb

```diff
@@ -1,49 +1,92 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n)
+# Space complexity: o(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  return 1 if n == 0
+  raise ArgumentError if n < 0
+
+  return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: o(n)
+# Space complexity: o(n)
 def reverse(s)
```

Comment:
👍 
This works, but because you create a new array with each recursive call this is O(n<sup>2</sup>) for both time/space complexity.